integration_test/integration_test.go

```diff
@@ -298,6 +350,783 @@ func (t *WASMIntegrationTestSuite) Test04_ContractExecution() {
+					proposalContent,
+					t.VolunteerValidatorWallet2,
+					[]*WalletInfo{t.ValidatorWallet1, t.ValidatorWallet2, t.ValidatorWallet3, t.ValidatorWallet4},
+				)
+
+				if err != nil {
+					return err
+				}
+
+				return nil
+			})
+
+			// slight delay for avoiding sequence mismatch
+			time.Sleep(time.Second / 5)
+		}
+
+		err := eg.Wait()
+
+		if assert.NoError(t.T(), err) {
+			fmt.Println("Proposal successfully applied!")
+		} else {
+			fmt.Println("Error detected on the proposal")
+			t.T().Fail()
+		}
+	}
+
+	fmt.Println("Waiting some blocks for the proposal passing...")
+	time.Sleep(time.Second*blocktime*validatorActiveBlocks + 1)
+
+	{
+		fmt.Println("Check existence of the volunteer validator")
+
+		client := volunteerValType.NewQueryClient(desc.GetConnectionWithContext(context.Background()))
+		validatorStatus, err := client.VolunteerValidators(context.Background(), &volunteerValType.QueryVolunteerValidatorsRequest{})
+		assert.NoError(t.T(), err)
+
+		thisVolunteerValAddress := sdktypes.ValAddress(t.VolunteerValidatorWallet2.ByteAddress).String()
+
+		if assert.NotContains(t.T(), validatorStatus.GetVolunteerValidators(), thisVolunteerValAddress) {
+			fmt.Println(thisVolunteerValAddress, "successfully removed from the validator set!")
+		} else {
+			fmt.Println(thisVolunteerValAddress, "still exist!")
+			t.T().Fail()
+		}
+	}
+}
+
+func (t *WASMIntegrationTestSuite) Test14_TryChangingGeneralValidatorToVolunteerValidator_ShouldFail() {
+	amt := sdktypes.NewInt(1_000000_000000_000000)
+
+	{
+		fmt.Println("Try registering as a volunteer validator from the general validator...")
+
+		proposalContent, err := volunteerValType.NewRegisterVolunteerValidatorProposal(
+			"register_volunteer_validator_from_general_validator",
+			"Test volunteer validator registary",
+			sdktypes.AccAddress(t.ValidatorWallet1.ByteAddress.Bytes()),
+			sdktypes.ValAddress(t.ValidatorWallet1.ByteAddress.Bytes()),
+			&ed25519.PubKey{Key: t.Validator1PVKey.PubKey.Bytes()},
+			sdktypes.NewCoin(xplatypes.DefaultDenom, amt), // smaller amount than other basic validators
+			stakingtype.NewDescription("volunteer_validator", "", "", "", ""),
+		)
+
+		assert.NoError(t.T(), err)
+
+		err = applyVoteTallyingProposal(
+			desc.GetConnectionWithContext(context.Background()),
+			proposalContent,
+			t.VolunteerValidatorWallet2,
+			[]*WalletInfo{t.ValidatorWallet1, t.ValidatorWallet2, t.ValidatorWallet3, t.ValidatorWallet4},
+		)
+
+		if assert.Error(t.T(), err) {
+			fmt.Println(err)
+			fmt.Println("Expected failure! Test succeeded!")
+		} else {
+			fmt.Println("Unexpected situation. Failure")
+			t.T().Fail()
+		}
+	}
+}
+
+func (t *WASMIntegrationTestSuite) Test15_ValidatorActiveSetChange() {
+	volunteerValDelegationAmt := sdktypes.NewInt(3_000000_000000_000000)
+	generalValUpperDelegationAmt := sdktypes.NewInt(4_000000_000000_000000)
+	generalValLowerDelegationAmt := sdktypes.NewInt(2_000000_000000_000000)
```

Comment:
Tendermint's power sort is based on integers. That is, 2_000000_000000_000000 is 2. If 3_000000_000000_000000 is slashed, it becomes 2_950000_000000_000000, which is also 2.

This is what causes the problem.

Reply:
https://github.com/xpladev/xpla/pull/84/commits/3b841f68bb36d0923043fb5501612a33a74a7f91